Add database deployment for Slurm accounting. (#452)
aws-samples/awsome-distributed-training: 1.architectures/8.accounting-database/README.md

# Deploy an accounting database for Slurm

This sample setup a Amazon RDS database for HPC and Machine Learning cluster.
You can use for Slurm accounting and generate report of your cluster usage.
For more information you can visit Slurm documentation on [accounting](https://slurm.schedmd.com/accounting.html)

## Deploy

> 1. Download the `cf_database-accounting.yaml` file

Reviewer: Add a 1-click deploy link like so:

```suggestion
1. Deploy the accounting database using the 1-click deploy:

[<kbd> <br> 1-Click Deploy 🚀 <br> </kbd>](https://console.aws.amazon.com/cloudformation/home?#/stacks/quickcreate?templateURL=https%3A%2F%2Fawsome-distributed-training.s3.amazonaws.com%2Ftemplates%2Fcf_database-accounting.yaml&stackName=accounting-database)

```

	  --output text)
	```

### Get the Database password
The password to access the database was generated randomly and stored in AWS Secret Manager under `AccountingClusterAdminSecre-XXX` output of the cloudformation stack.

1. Get the Secret ARN
	```bash
	 DATABASE_SECRET_ARN=$(aws cloudformation describe-stacks \
	 	--stack-name slurm-accounting-database \
	  --query 'Stacks[0].Outputs[?OutputKey==`DatabaseSecretArn`].OutputValue' \
	  --output text)
	```

## Configure AWS ParallelCluster
Starting with version 3.3.0, AWS ParallelCluster supports Slurm accounting with the cluster configuration parameter `SlurmSettings / Database`.

To use the database created previously for accounting, add the following in the `SlurmSettings` section of your cluster configuration file:

```yaml
    Database:
      Uri: ${DATABASE_URI}:3306
      UserName: ${DATABASE_ADMIN}
      PasswordSecretArn: ${DATABASE_SECRET_ARN}
    CustomSlurmSettings:
      # Enable accounting for GPU resources.
      # - AccountingStorageTRES: gres/gpu
      - AccountingStorageTRES: gres/gpu
```

## Amazon SageMaker HyperPod Orchestrated by Slurm
There are 3 steps to setup Slurm with the accounting database:

> 	1. sdd

Reviewer: fill out steps.

Author: sorry. good catch.

# Deploy an accounting database for Slurm

This sample setup a Amazon RDS database for HPC and Machine Learning cluster.
You can use for Slurm accounting and generate report of your cluster usage.
For more information you can visit Slurm documentation on [accounting](https://slurm.schedmd.com/accounting.html)

> 

Reviewer 2: ```suggestion
### Prerequisites

You need to have a Slurm cluster deployed on VPC with at least two private subnets.
```

## Deploy

Deploy the accounting database using the 1-click deploy:

> [<kbd> <br> 1-Click Deploy 🚀 <br> </kbd>](https://console.aws.amazon.com/cloudformation/home?#/stacks/quickcreate?templateURL=https%3A%2F%2Fawsome-distributed-training.s3.amazonaws.com%2Ftemplates%2Fcf_database-accounting.yaml&stackName=accounting-database)

Reviewer 2: ```suggestion
[<kbd> <br> 1-Click Deploy 🚀 <br> </kbd>](https://console.aws.amazon.com/cloudformation/home?#/stacks/quickcreate?templateURL=https%3A%2F%2Fawsome-distributed-training.s3.amazonaws.com%2Ftemplates%2Fcf_database-accounting.yaml&stackName=slurm-accounting-database)
```
better to make the default name consistent.

**Note** or you can deploy using AWS cli and CloudFormation:
  ```bash

>   aws cloudformation deploy --stack-name slurm-accounting-database \

Reviewer 2: ```suggestion
  aws cloudformation deploy --stack-name accounting-database \
```
better to make the default name consistent.

```

## Configure AWS ParallelCluster
Starting with version 3.3.0, AWS ParallelCluster supports Slurm accounting with the cluster configuration parameter `SlurmSettings / Database`.

To use the database created previously for accounting, add the following in the `SlurmSettings` section of your cluster configuration file:

```yaml
    Database:
      Uri: ${DATABASE_URI}:3306
      UserName: ${DATABASE_ADMIN}
      PasswordSecretArn: ${DATABASE_SECRET_ARN}
    CustomSlurmSettings:
      # Enable accounting for GPU resources.
      # - AccountingStorageTRES: gres/gpu
      - AccountingStorageTRES: gres/gpu
```

## Amazon SageMaker HyperPod Orchestrated by Slurm
There are two steps to setup Slurm with the accounting database:
1. Add database configuration file
1. Configure Slurm accounting

### Add database configuration file
You need to execute the following command on the controller node to configure the database connectivity for Slurm.

```bash
cat > /opt/slurm/etc/slurmdbd.conf << EOF
AuthType=auth/munge
DbdHost=$(hostname) # Slurm controller ip address.
DbdPort=6819
SlurmUser=slurm
LogFile=/var/log/slurmdbd.log
StorageType=accounting_storage/mysql
StorageUser=${DATABASE_ADMIN}

> StoragePass=$(aws secretsmanager get-secret-value --secret-id ${DATABASE_SECRET_ARN} --query SecretString --output text)

Reviewer 2: Is the `StoragePass` static? By default, role for SMHP headnode does not have permission to run those command, so we need
1/ create the file locally
2/ add necessary permission to the role.
If the `StoragePass` is not static we need to have a description for 2/.

Author: yes it is the password. Will update. Thank you.

StorageHost=${DATABASE_URI}
StoragePort=3306
EOF
```

### Configure Slurm accounting

```bash
cat >> /opt/slurm/etc/slurm.conf << EOF
# ACCOUNTING
JobAcctGatherType=jobacct_gather/linux
JobAcctGatherFrequency=60
AccountingStorageType=accounting_storage/slurmdbd
AccountingStorageHost=$(hostname)
AccountingStorageUser=${DATABASE_ADMIN}
AccountingStoragePort=6819

> AccountingStorageTRES: gres/gpu

Reviewer 2: I got 
```
scontrol: error: Parse error in file /opt/slurm/etc/slurm.conf line 76: "AccountingStorageTRES: gres/gpu"
scontrol: fatal: Unable to process configuration file
```

Reviewer 2: with the following `slurm.conf`:
```
# slurm.conf file generated by AWS

ClusterName=ml-cluster

SlurmctldHost=ip-10-1-28-102(10.1.28.102)
SlurmctldParameters=enable_configless

ProctrackType=proctrack/linuxproc
RebootProgram="/usr/bin/systemctl reboot"
ReturnToService=1

# Slurmctld settings
SlurmctldLogFile=/var/log/slurm/slurmctld.log
StateSaveLocation=/var/spool/slurmctld

# Slurmd settings
SlurmdLogFile=/var/log/slurm/slurmd.log
SlurmdSpoolDir=/var/spool/slurmd

# Allow dynamic node registration
TreeWidth=65533
MaxNodeCount=32768

SlurmUser=slurm
AuthType=auth/munge

SwitchType=switch/none
TaskPlugin=task/none # Required for auto-resume feature.

# TIMERS
InactiveLimit=0
KillWait=30
MinJobAge=300
ResumeTimeout=300
SlurmctldTimeout=120
SlurmdTimeout=300
UnkillableStepTimeout=300
WaitTime=0

# SCHEDULING
SchedulerType=sched/backfill
SelectType=select/cons_tres

LaunchParameters=enable_nss_slurm
SchedulerParameters=permit_job_expansion # Required for auto-resume feature.

# ACCOUNTING
Include accounting.conf

# There is no nodes in this cluster yet
PartitionName=dev Nodes=ALL Default=YES MaxTime=INFINITE OverSubscribe=EXCLUSIVE State=UP
PartitionName=p5.48xlarge Nodes=ip-10-1-66-71,ip-10-1-34-241,ip-10-1-108-157,ip-10-1-43-135,ip-10-1-85-132,ip-10-1-94-217,ip-10-1-25-80,ip-10-1-18-140,ip-10-1-42-99,ip-10-1-82-214,ip-10-1-98-125,ip-10-1-30-217,ip-10-1-53-9,ip-10-1-2-238,ip-10-1-82-85,ip-10-1-104-48

NodeName=ip-10-1-66-71 NodeHostname=ip-10-1-66-71 NodeAddr=10.1.66.71 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-34-241 NodeHostname=ip-10-1-34-241 NodeAddr=10.1.34.241 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-108-157 NodeHostname=ip-10-1-108-157 NodeAddr=10.1.108.157 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-43-135 NodeHostname=ip-10-1-43-135 NodeAddr=10.1.43.135 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-85-132 NodeHostname=ip-10-1-85-132 NodeAddr=10.1.85.132 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-94-217 NodeHostname=ip-10-1-94-217 NodeAddr=10.1.94.217 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-25-80 NodeHostname=ip-10-1-25-80 NodeAddr=10.1.25.80 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-18-140 NodeHostname=ip-10-1-18-140 NodeAddr=10.1.18.140 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-42-99 NodeHostname=ip-10-1-42-99 NodeAddr=10.1.42.99 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-82-214 NodeHostname=ip-10-1-82-214 NodeAddr=10.1.82.214 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-98-125 NodeHostname=ip-10-1-98-125 NodeAddr=10.1.98.125 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-30-217 NodeHostname=ip-10-1-30-217 NodeAddr=10.1.30.217 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-53-9 NodeHostname=ip-10-1-53-9 NodeAddr=10.1.53.9 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-2-238 NodeHostname=ip-10-1-2-238 NodeAddr=10.1.2.238 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-82-85 NodeHostname=ip-10-1-82-85 NodeAddr=10.1.82.85 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1
NodeName=ip-10-1-104-48 NodeHostname=ip-10-1-104-48 NodeAddr=10.1.104.48 Boards=1 CPUs=96 CoresPerSocket=48 Gres=gpu:h100:8 RealMemory=2097152 SocketsPerBoard=2 State=CLOUD ThreadsPerCore=1# ACCOUNTING
JobAcctGatherType=jobacct_gather/linux
JobAcctGatherFrequency=60
AccountingStorageType=accounting_storage/slurmdbd
AccountingStorageHost=ip-10-1-28-102
AccountingStorageUser=clusteradmin
AccountingStoragePort=6819
AccountingStorageTRES: gres/gpu
```

Author: Changed. Thank you

> AccountingStorageTRES: gres/gpu

Reviewer 2: ```suggestion
AccountingStorageTRES=gres/gpu
```

You can use for Slurm accounting and generate report of your cluster usage.
For more information you can visit Slurm documentation on [accounting](https://slurm.schedmd.com/accounting.html).

You will need at least two private subnets in different avaibility zones to deploy the database.

## Deploy

Deploy the accounting database using the 1-click deploy:

[<kbd> <br> 1-Click Deploy 🚀 <br> </kbd>](https://console.aws.amazon.com/cloudformation/home?#/stacks/quickcreate?templateURL=https%3A%2F%2Fawsome-distributed-training.s3.amazonaws.com%2Ftemplates%2Fcf_database-accounting.yaml&stackName=slurm-accounting-database)

**Note** or you can deploy using AWS cli and CloudFormation:
  ```bash
  aws cloudformation deploy --stack-name slurm-accounting-database \
  --template-file cf_database-accounting.yaml \
  --capabilities CAPABILITY_IAM CAPABILITY_NAMED_IAM CAPABILITY_AUTO_EXPAND \
  --parameter-overrides VpcId=XXX SubnetIds=XXX,XXX
  ```

## Get database parameters
In this section, you will retrieve the database parameter that are used by Slurm to connect to the accounting database.

### Get the Database URI

> Retrieve the `DatabaseHost` to connect to the LDAP User Interface.

Reviewer 2: I am not quite sure 1/ how LDAP user interface related in this context 2/  why we need to "Copy URL into a Web Browser". Could you please elavorate?

Get the Secret ARN
```bash
DATABASE_SECRET_ARN=$(aws cloudformation describe-stacks \
  --stack-name slurm-accounting-database \
  --query 'Stacks[0].Outputs[?OutputKey==`DatabaseSecretArn`].OutputValue' \
  --output text)
```

## Configure AWS ParallelCluster
Starting with version 3.3.0, AWS ParallelCluster supports Slurm accounting with the cluster configuration parameter `SlurmSettings / Database`.

To use the database created previously for accounting, add the following in the `SlurmSettings` section of your cluster configuration file:

```yaml
    Database:
      Uri: ${DATABASE_URI}:3306
      UserName: ${DATABASE_ADMIN}
      PasswordSecretArn: ${DATABASE_SECRET_ARN}
    CustomSlurmSettings:
      # Enable accounting for GPU resources.
      # - AccountingStorageTRES: gres/gpu
      - AccountingStorageTRES: gres/gpu
```

## Amazon SageMaker HyperPod Orchestrated by Slurm
There are two steps to setup Slurm with the accounting database:
1. Add database configuration file
1. Configure Slurm accounting

### Add database configuration file

> You need to execute the following command on the controller node to configure the database connectivity for Slurm.

Reviewer 2: ```suggestion
You need to execute the following command on the local terminal and place them into `/opt/slurm/etc` to configure the database connectivity for Slurm.
```

```bash

> cat > /opt/slurm/etc/slurmdbd.conf << EOF

Reviewer 2: ```suggestion
# Place the following file in /opt/slurm/etc/slurmdbd.conf 
```

AuthType=auth/munge
DbdHost=$(hostname) # Slurm controller ip address.
DbdPort=6819
SlurmUser=slurm
LogFile=/var/log/slurmdbd.log
StorageType=accounting_storage/mysql
StorageUser=${DATABASE_ADMIN}
StoragePass=$(aws secretsmanager get-secret-value --secret-id ${DATABASE_SECRET_ARN} --query SecretString --output text)
StorageHost=${DATABASE_URI}
StoragePort=3306

> EOF

Reviewer 2: ```suggestion
```

```

### Configure Slurm accounting

```bash

> cat >> /opt/slurm/etc/slurm.conf << EOF

Reviewer 2: # Append the following lines in /opt/slurm/etc/slurm.conf

> cat >> /opt/slurm/etc/slurm.conf << EOF

Reviewer 2: ```suggestion
# Append the following lines in /opt/slurm/etc/slurm.conf
```

# ACCOUNTING
JobAcctGatherType=jobacct_gather/linux
JobAcctGatherFrequency=60
AccountingStorageType=accounting_storage/slurmdbd
AccountingStorageHost=$(hostname)
AccountingStorageUser=${DATABASE_ADMIN}
AccountingStoragePort=6819
AccountingStorageTRES=gres/gpu

> EOF

Reviewer 2: ```suggestion
```

```

Restart the slurmctld to pickup the configuration change.
```bash

> sudo systemctl restart slurmdctld

Reviewer 2: ```suggestion
sudo systemctl restart slurmdctld.service
```

> sudo scontrol reconfigure

Reviewer 2: After this command I got
```
slurm_reconfigure error: Socket timed out on send/recv operation
```

```
ubuntu@ip-10-1-13-33:~$ sudo systemctl status slurmctld.service 
● slurmctld.service - Slurm controller daemon
     Loaded: loaded (/etc/systemd/system/slurmctld.service; enabled; vendor preset: enabled)
     Active: active (running) since Sat 2024-10-12 13:05:45 UTC; 3min 56s ago
       Docs: man:slurmctld(8)
   Main PID: 99908 (slurmctld)
      Tasks: 20
     Memory: 42.5M
     CGroup: /system.slice/slurmctld.service
             ├─ 99908 /opt/slurm/sbin/slurmctld -D
             └─100134 slurmctld: slurmscriptd

Oct 12 13:07:45 ip-10-1-13-33 slurmctld[99908]: slurmctld: error: "Include" failed in file /opt/slurm/etc/plugstack.conf line 2
Oct 12 13:07:45 ip-10-1-13-33 slurmctld[99908]: slurmctld: Running as primary controller
Oct 12 13:07:45 ip-10-1-13-33 slurmctld[99908]: slurmctld: error: s_p_parse_file: cannot stat file /opt/slurm/etc/plugstack.conf.d/*: No such file or directory, retrying in 1sec up to 60sec
Oct 12 13:07:45 ip-10-1-13-33 slurmctld[99908]: slurmctld: error: slurm_send_node_msg: [socket:[760262]] slurm_bufs_sendto(msg_type=RESPONSE_PARTITION_INFO) failed: Unexpected missing socket error
Oct 12 13:07:45 ip-10-1-13-33 slurmctld[99908]: slurmctld: error: slurm_send_node_msg: [socket:[760263]] slurm_bufs_sendto(msg_type=RESPONSE_PARTITION_INFO) failed: Unexpected missing socket error
Oct 12 13:07:45 ip-10-1-13-33 slurmctld[99908]: slurmctld: error: slurm_send_node_msg: [socket:[760264]] slurm_bufs_sendto(msg_type=RESPONSE_PARTITION_INFO) failed: Unexpected missing socket error
Oct 12 13:07:45 ip-10-1-13-33 slurmctld[99908]: slurmctld: error: slurm_send_node_msg: [socket:[760266]] slurm_bufs_sendto(msg_type=RESPONSE_STATS_INFO) failed: Unexpected missing socket error
Oct 12 13:07:45 ip-10-1-13-33 slurmctld[99908]: slurmctld: error: slurm_send_node_msg: [socket:[760265]] slurm_bufs_sendto(msg_type=RESPONSE_PARTITION_INFO) failed: Unexpected missing socket error
Oct 12 13:07:45 ip-10-1-13-33 slurmctld[99908]: slurmctld: error: slurm_send_node_msg: [socket:[760261]] slurm_bufs_sendto(msg_type=RESPONSE_NODE_INFO) failed: Unexpected missing socket error
Oct 12 13:08:44 ip-10-1-13-33 slurmctld[99908]: slurmctld: error: "Include" failed in file /opt/slurm/etc/plugstack.conf line 2
ubuntu@ip-10-1-13-33:~$ sudo vim /opt/slurm/etc/plugstack.conf 
```
